File: java/src/test/java/org/taymyr/lagom/elasticsearch/search/ElasticSearchIT.java
package org.taymyr.lagom.elasticsearch.search;

import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.pcollections.HashTreePMap;
import org.taymyr.lagom.elasticsearch.AbstractElasticsearchIT;
import org.taymyr.lagom.elasticsearch.TestDocument;
import org.taymyr.lagom.elasticsearch.TestDocument.TestDocumentResult;
import org.taymyr.lagom.elasticsearch.document.dsl.index.IndexResult;
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex;
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex.Settings;
import org.taymyr.lagom.elasticsearch.indices.dsl.Mapping;
import org.taymyr.lagom.elasticsearch.indices.dsl.MappingProperty;
import org.taymyr.lagom.elasticsearch.search.dsl.SearchRequest;
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.ExistsQuery;
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.IdsQuery;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;
import java.util.stream.IntStream;

import static java.lang.Thread.sleep;
import static java.time.temporal.ChronoUnit.SECONDS;
import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.tuple;
import static org.taymyr.lagom.elasticsearch.ServiceCall.invoke;
import static org.taymyr.lagom.elasticsearch.indices.dsl.DataType.KEYWORD;
import static org.taymyr.lagom.elasticsearch.indices.dsl.DataType.TEXT;
import static org.taymyr.lagom.elasticsearch.search.ScrollSearchSourceStage.scrollSearchSource;
import static org.taymyr.lagom.elasticsearch.search.dsl.query.Order.asc;

class ElasticSearchIT extends AbstractElasticsearchIT {

    @Test
    @DisplayName("Search service descriptor should work correct")
    void shouldWorkCorrect() throws InterruptedException, ExecutionException, TimeoutException {
        createDocument();
        sleep(1000);
        searchById();
    }

    void createDocument() throws InterruptedException, ExecutionException, TimeoutException {
        IndexResult result = eventually(invoke(elasticDocument.indexWithId("test", "1"),
                new TestDocument("user", "message")));
        assertThat(result.getIndex()).isEqualTo("test");
        assertThat(result.getType()).isEqualTo("_doc");
    }

    void check(TestDocumentResult result) {
        assertThat(result.getTamedOut()).isFalse();
        assertThat(result.getHits().getTotal().getValue()).isEqualTo(1);
        assertThat(result.getHits().getHits()).hasSize(1);
        assertThat(result.getHits().getHits().get(0).getScore()).isEqualTo(1.0);
        assertThat(result.getHits().getHits().get(0).getSource().getUser()).isEqualTo("user");
        assertThat(result.getHits().getHits().get(0).getSource().getMessage()).isEqualTo("message");
    }

    void searchById() throws InterruptedException, ExecutionException, TimeoutException {
        SearchRequest searchRequest = new SearchRequest(IdsQuery.of("1"));
        check(eventually(invoke(elasticSearch.search(asList("test")), searchRequest, TestDocumentResult.class)));
        check(eventually(invoke(elasticSearch.search(asList("test")), searchRequest, TestDocumentResult.class)));
        check(eventually(invoke(elasticSearch.search("test"), searchRequest, TestDocumentResult.class)));
        check(eventually(invoke(elasticSearch.search("test"), searchRequest, TestDocumentResult.class)));
        assertThat(eventually(elasticSearch.count("test").invoke(searchRequest)).getCount()).isEqualTo(1);
    }

    @Test
    void testScrollingSearch() throws Throwable {
        String indexName = "search-scroller-idx";
        elasticIndices.create(indexName).invoke(new CreateIndex(
            new Settings(1, 1),
            new Mapping(
                HashTreePMap.<String, MappingProperty>empty()
                    .plus("user", MappingProperty.builder().type(KEYWORD).build())
                    .plus("message", MappingProperty.builder().type(TEXT).build())
            )
        ));
        sleep(Duration.of(5, SECONDS).toMillis());
        int firstId = 1;
        int latestId = 12000;
        IntStream.range(firstId, latestId + 1).forEach(i -> {
            try {
                eventually(invoke(elasticDocument.indexWithId(indexName, String.valueOf(i)),
                    new TestDocument("user" + i, "message" + i)));
            } catch (Throwable e) {
                throw  new RuntimeException(e);
            }
        });
        sleep(Duration.of(5, SECONDS).toMillis());
        SearchRequest searchRequest = SearchRequest.builder()
            .query(ExistsQuery.of("user"))
            .sort(asc("user"))
            .size(1000)
            .build();
        List<TestDocument> foundDocs;
        foundDocs = loggingTimings(
            "Search using " + SearchScroller.class.getName(),
            () -> eventually(
                new SearchScroller(elasticSearch, indexName).searchAfter(searchRequest, TestDocumentResult.class),
                Duration.of(30, SECONDS)
            )
        );
        assertThat(foundDocs).isNotEmpty().hasSize(latestId)
            .extracting(TestDocument::getUser, TestDocument::getMessage)
            .contains(
                tuple("user" + latestId, "message" + latestId),
                tuple("user" + firstId, "message" + firstId)
            );
        foundDocs = loggingTimings(
            "Search using " + ScrollSearchSourceStage.class.getName(),
            () -> eventually(
                scrollSearchSource(searchRequest, request -> invoke(elasticSearch.search(indexName), request, TestDocumentResult.class))
                    .mapAsyncUnordered(2, d ->
                        CompletableFuture.supplyAsync(() -> {
                            try {
                                return d.getSources();
                            } catch (Throwable throwable) {
                                throw new RuntimeException(throwable);
                            }
                        })
                    )
                    .toMat(
                        Sink.reduce((next, acc) -> {
                            acc.addAll(next);
                            return acc;
                        }),
                        Keep.right()
                    )
                    .run(materializer),
                Duration.of(30, SECONDS)
            )
        );
        assertThat(foundDocs).isNotEmpty().hasSize(latestId)
            .extracting(TestDocument::getUser, TestDocument::getMessage)
            .contains(
                tuple("user" + latestId, "message" + latestId),
                tuple("user" + firstId, "message" + firstId)
            );
    }
}

File: src/test/kotlin/org/taymyr/lagom/elasticsearch/search/ElasticSearchIT.kt
package org.taymyr.lagom.elasticsearch.search

import io.kotlintest.extensions.TestListener
import io.kotlintest.matchers.beInstanceOf
import io.kotlintest.matchers.collections.shouldHaveSize
import io.kotlintest.shouldBe
import io.kotlintest.shouldNotBe
import io.kotlintest.specs.WordSpec
import io.kotlintest.whenReady
import org.taymyr.lagom.elasticsearch.AutocompleteFilter
import org.taymyr.lagom.elasticsearch.IndexedSampleCategory
import org.taymyr.lagom.elasticsearch.LagomClientAndEmbeddedElastic
import org.taymyr.lagom.elasticsearch.MessageKeywordTerm
import org.taymyr.lagom.elasticsearch.SampleCategory
import org.taymyr.lagom.elasticsearch.SampleCategoryResult
import org.taymyr.lagom.elasticsearch.SampleDocument
import org.taymyr.lagom.elasticsearch.SampleDocumentResult
import org.taymyr.lagom.elasticsearch.UserKeywordTerm
import org.taymyr.lagom.elasticsearch.deser.invoke
import org.taymyr.lagom.elasticsearch.deser.invokeT
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkCreate
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkRequest
import org.taymyr.lagom.elasticsearch.document.dsl.bulk.BulkResult
import org.taymyr.lagom.elasticsearch.indices.dsl.CreateIndex
import org.taymyr.lagom.elasticsearch.indices.dsl.CustomAnalyzer
import org.taymyr.lagom.elasticsearch.indices.dsl.DataType
import org.taymyr.lagom.elasticsearch.indices.dsl.Mapping
import org.taymyr.lagom.elasticsearch.indices.dsl.MappingProperty
import org.taymyr.lagom.elasticsearch.search.dsl.SearchRequest
import org.taymyr.lagom.elasticsearch.search.dsl.query.compound.BoolQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.Match
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.MatchQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.fulltext.MultiMatchQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.Ids
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.IdsQuery
import org.taymyr.lagom.elasticsearch.search.dsl.query.term.TermQuery
import java.lang.Thread.sleep

class ElasticSearchIT : WordSpec() {

    override fun listeners(): List<TestListener> = listOf(Companion)

    init {
        "ElasticSearch" should {
            "put documents" {
                repeat(10) {
                    whenReady(elasticDocument.indexWithId("test", "sample", it.toString())
                        .invoke(SampleDocument("user-$it", "message-$it")).toCompletableFuture()) { result ->
                        result.index shouldBe "test"
                        result.type shouldBe "sample"
                    }
                }
            }
            "search a document by id" {
                sleep(1000)
                val searchRequest = SearchRequest(IdsQuery(Ids(values = listOf("2"))))
                val check = { result: SampleDocumentResult ->
                    result.tamedOut shouldBe false
                    result.hits.run {
                        total shouldBe 1
                        hits shouldHaveSize 1
                        hits[0].score shouldBe 1.0
                        hits[0].source.user shouldBe "user-2"
                        hits[0].source.message shouldBe "message-2"
                    }
                }
                whenReady(elasticSearch.search(listOf("test"), listOf("sample")).invokeT<SearchRequest, SampleDocumentResult>(searchRequest).toCompletableFuture(), check)
                whenReady(elasticSearch.search(listOf("test")).invokeT<SearchRequest, SampleDocumentResult>(searchRequest).toCompletableFuture(), check)
                whenReady(elasticSearch.search("test", "sample").invokeT<SearchRequest, SampleDocumentResult>(searchRequest).toCompletableFuture(), check)
                whenReady(elasticSearch.search("test").invokeT<SearchRequest, SampleDocumentResult>(searchRequest).toCompletableFuture(), check)
                whenReady(elasticSearch.search().invokeT<SearchRequest, SampleDocumentResult>(searchRequest).toCompletableFuture(), check)
            }
        }
        "Advanced ElasticSearch" should {
            "create index with analyzer" {
                val request = CreateIndex(
                    CreateIndex.Settings(1, 1, CreateIndex.Analysis(
                        mapOf(
                            "autocomplete_filter" to AutocompleteFilter(
                                "edge_ngram",
                                1,
                                20
                            )
                        ),
                        mapOf(
                            "autocomplete" to CustomAnalyzer(
                                "standard",
                                listOf(
                                    "lowercase",
                                    "autocomplete_filter"
                                )
                            )
                        )
                    )),
                    mapOf(
                        "some_type" to Mapping(mapOf(
                            "id" to MappingProperty.LONG,
                            "name" to MappingProperty(type = DataType.TEXT, analyzer = "autocomplete"),
                            "title" to MappingProperty.OBJECT,
                            "technicalName" to MappingProperty.TEXT,
                            "fullText" to MappingProperty.TEXT,
                            "fullTextBoosted" to MappingProperty.TEXT,
                            "attachAllowed" to MappingProperty.BOOLEAN
                        ))
                    )
                )
                whenReady(elasticIndices.create("category").invoke(request).toCompletableFuture()) { result ->
                    result.acknowledged shouldBe true
                    result.shardsAcknowledged shouldBe true
                    result.index shouldBe "category"
                }
            }
            "successful add test category" {
                val request = BulkRequest.ofCommands(
                    BulkCreate("1",
                        IndexedSampleCategory(
                            SampleCategory(
                                1,
                                listOf("test1"),
                                null,
                                null,
                                true,
                                "Овощи",
                                "Огурцы"
                            )
                        )
                    ),
                    BulkCreate("2",
                        IndexedSampleCategory(
                            SampleCategory(
                                1,
                                listOf("test1"),
                                null,
                                null,
                                true,
                                "Овощи",
                                "Капуста"
                            )
                        )
                    )
                )
                whenReady(elasticDocument.bulk("category", "some_type").invoke(request).toCompletableFuture()) { result ->
                    result shouldBe beInstanceOf(BulkResult::class)
                    result.errors shouldBe false
                    result.items shouldHaveSize 2
                    result.items[0].run {
                        this.status shouldBe 201
                        this.result shouldBe "created"
                        this.error shouldBe null
                        this.index shouldBe "category"
                        this.type shouldBe "some_type"
                        this.id shouldBe "1"
                    }
                }
            }
            "successful search document using autocomplete filter" {
                sleep(1000)
                val searchRequest = SearchRequest(
                    MatchQuery(object : Match {
                        val name = "test1"
                    })
                )
                whenReady(elasticSearch.search(listOf("category"), listOf("some_type"))
                    .invokeT<SearchRequest, SampleCategoryResult>(searchRequest).toCompletableFuture()) { result ->
                    result.tamedOut shouldBe false
                    result.hits.run {
                        total shouldBe 2
                        hits shouldHaveSize 2
                        hits[0].source.name shouldNotBe null
                        hits[0].source.name!![0] shouldBe "test1"
                    }
                }
            }
            "successful search document using multi-match" {
                sleep(1000)
                val searchRequest = SearchRequest(
                    MultiMatchQuery.of(
                        "Огурцовые овощи",
                        mapOf(
                            Pair("fullTextBoosted", 10),
                            Pair("fullText", 3)
                        )
                    )
                )
                whenReady(elasticSearch.search(listOf("category"), listOf("some_type"))
                    .invokeT<SearchRequest, SampleCategoryResult>(searchRequest).toCompletableFuture()) { result ->
                    result.tamedOut shouldBe false
                    result.hits.run {
                        total shouldBe 2
                        hits shouldHaveSize 2
                        hits[0].source.name shouldNotBe null
                        hits[0].source.name!![0] shouldBe "test1"
                    }
                }
            }
        }
        "Request Body Search with Query DSL" should {
            "successfully perform BoolQuery search" {
                val userTerm = UserKeywordTerm("user-9")
                val messageTerm = MessageKeywordTerm("message-9")
                val request = SearchRequest(
                    BoolQuery.boolQuery()
                        .must(listOf(TermQuery.ofTerm(userTerm), TermQuery.ofTerm(messageTerm)))
                        .build(),
                    0, 100
                )
                whenReady(elasticSearch.search(listOf("test"), listOf("sample")).invokeT<SearchRequest, SampleDocumentResult>(request)
                    .toCompletableFuture()) {
                    val found = it.hits.hits.filter {
                        h -> h.source.user == userTerm.user && h.source.message == messageTerm.message
                    }
                    found shouldHaveSize 1
                }
            }
        }
    }

    companion object : LagomClientAndEmbeddedElastic()
}